Review Feedback from Mohamed Boucadair (#52)
suit-wg/suit-mti: draft-ietf-suit-mti.md

---

> title: Cryptographic Algorithms for Internet of Things Devices

Reviewer: ```suggestion
title: Cryptographic Algorithms for Internet of Things (IoT) Devices
```

informative:
  I-D.ietf-suit-firmware-encryption:
  RFC9053:
  RFC9019:
  RFC6973:
  I-D.ietf-teep-protocol:
  IANA-COSE:
    title: "CBOR Object Signing and Encryption (COSE)"
    author:
    date: 2022
    target: https://www.iana.org/assignments/cose/cose.xhtml
  LwM2M:
    target: "https://www.openmobilealliance.org/specifications/lwm2m"
    title: "OMA Lightweight M2M"
    date: 2025-04-20
    author:
    - org: "Open Mobile Alliance"

--- abstract

This document defines cryptographic algorithm profiles for use with the Software Updates for Internet
of Things (SUIT) manifest. These profiles specify sets of algorithms to promote interoperability across
implementations.

> The SUIT manifest, as defined in RFC 9124, provides a flexible and extensible format for describing how

Reviewer: ```suggestion
The SUIT manifest, as defined in "A Manifest Information Model for Firmware Updates in Internet of Things (IoT) Devices" (RFC 9124), provides a flexible and extensible format for describing how
```

Reviewer: The abstract should be self contained.

firmware and software updates are to be fetched, verified, decrypted, and installed on resource-constrained
devices. To ensure the security of these update processes, the manifest relies on cryptographic algorithms
for functions such as digital signature verification, integrity checking, and confidentiality.

Given the diversity of IoT deployments and the evolving cryptographic landscape, algorithm agility is
essential. This document groups algorithms into named profiles to accommodate varying levels of device
capabilities and security requirements. These profiles support the use cases laid out in the SUIT architecture,

> published with RFC 9019.

Reviewer: ```suggestion
published in "A Firmware Update Architecture for Internet of Things" (RFC 9019).
```

Reviewer: The abstract should be self-contained.

--- middle

#  Introduction

> This document defines algorithm profiles intended for authors of SUIT manifests and their recipients,

Reviewer: ```suggestion
This document defines algorithm profiles intended for authors of Software Updates of Internet of Things (SUIT) manifests and their recipients,
```

with the goal of promoting interoperability in software update scenarios for constrained nodes. These
profiles specify sets of algorithms that are tailored to the evolving security landscape, recognizing
that cryptographic requirements may change over time. 

The following profiles are defined:

* One profile designed for constrained devices that support only symmetric key cryptography
* Two profiles for constrained devices capable of using asymmetric key cryptography
* Two profiles that employ Authenticated Encryption with Associated Data (AEAD) ciphers
* One constrained asymmetric profile that uses a hash-based signature scheme

Due to the asymmetric nature of SUIT deployments - where manifest authors typically operate in
resource-rich environments while recipients are resource-constrained - the cryptographic
requirements differ between these two roles.

This specification uses AES-CTR in combination with a digest algorithm, as defined in {{-ctrcbc}},
to support use cases that require out-of-order block reception and decryption-capabilities not

> offered by AEAD algorithms. For further discussion of these constrained use cases, see

Reviewer: ```suggestion
offered by AEAD algorithms. For further discussion of these constrained use cases, refer to
```

The use of the profiles by authors and recipients is based on the following assumptions:

> - Recipients MAY choose which MTI profile they wish to implement. It is RECOMMENDED that they implement the `suit-sha256-hsslms-a256kw-a256ctr` profile. Recipients MAY implement any number of other profiles not defined in this document. Recipients MAY choose not to implement encryption and the corresponding key exchange algorithms if they do not intend to support encrypted payloads.

Reviewer: ```suggestion
- Recipients MAY choose which MTI profile they wish to implement. It is RECOMMENDED that they implement the `suit-sha256-hsslms-a256kw-a256ctr` profile ({{suit-sha256-hsslms-a256kw-a256ctr}}). Recipients MAY implement any number of other profiles not defined in this document. Recipients MAY choose not to implement encryption and the corresponding key exchange algorithms if they do not intend to support encrypted payloads.
```

Author: Looks good to me.

Author: For some reason I cannot merge this change.

The use of the profiles by authors and recipients is based on the following assumptions:

> - Recipients MAY choose which profile they wish to implement. It is RECOMMENDED that they implement the `suit-sha256-hsslms-a256kw-a256ctr` profile. Recipients MAY implement any number of other profiles not defined in this document. Recipients MAY choose not to implement encryption and the corresponding key exchange algorithms if they do not intend to support encrypted payloads.

Reviewer: ```suggestion
- Recipients MAY choose which profile they wish to implement. It is RECOMMENDED that they implement the `suit-sha256-hsslms-a256kw-a256ctr` profile ({{suit-sha256-hsslms-a256kw-a256ctr}}). Recipients MAY implement any number of other profiles not defined in this document. Recipients MAY choose not to implement encryption and the corresponding key exchange algorithms if they do not intend to support encrypted payloads.
```

| Key Exchange | ECDH-ES + A128KW | -29 |
| Encryption | A128GCM | 1 |

## Profile `suit-sha256-ed25519-ecdh-chacha-poly` {#suit-sha256-ed25519-ecdh-chacha-poly}

> This profile also supports asymmetric algorithms with AEAD algorithms but offers an alternative to `suit-sha256-esp256-ecdh-a128gcm`.

Reviewer: ```suggestion
This profile also supports asymmetric algorithms with AEAD algorithms but offers an alternative to `suit-sha256-esp256-ecdh-a128gcm` ({{suit-sha256-esp256-ecdh-a128gcm}}).
```

| Key Exchange | ECDH-ES + A128KW | -29 |
| Encryption | ChaCha20/Poly1305 | 24 |

## Profile `suit-sha256-hsslms-a256kw-a256ctr` {#suit-sha256-hsslms-a256kw-a256ctr}

> This profile utilzes a stateful hash-based signature algorithm, namely the Hierarchical Signature System / Leighton-Micali Signature (HSS/LMS), as a unique alternative to the  profiles listed above.

Reviewer: ```suggestion
This profile utilizes a stateful hash-based signature algorithm, namely the Hierarchical Signature System / Leighton-Micali Signature (HSS/LMS), as a unique alternative to the  profiles listed above.
```

A note regarding the use of the HSS/LMS: The decision as to how deep the tree is, is a decision that affects authoring tools only (see {{RFC8778}}).

> Verification is not affected by the choice of the "W" parameter, but the size of the signature is affected. To support the long lifetimes required by IoT devices, it is RECOMMENDED to use trees with greater height (see Section 2.2 of {{RFC8778}}).

Reviewer: ```suggestion
Verification is not affected by the choice of the "W" parameter, but the size of the signature is affected. To support the long lifetimes required by IoT devices, it is RECOMMENDED to use trees with greater height ({{Section 2.2 of RFC8778}}).
```

| Algorithm Type | Algorithm | COSE Key |
|============|
| Digest | SHA-256 | -16 |
| Authentication | HSS/LMS | -46 |
| Key Exchange | A256KW | -5 |
| Encryption | A256CTR | -65532 |

# Security Considerations {#security}

Payload encryption is used to protect sensitive content such as machine learning models, proprietary algorithms, and personal data {{RFC6973}}.
In the context of SUIT, the primary purpose of payload encryption is to defend against unauthorized observation during distribution. By encrypting
the payload, confidential information can be safeguarded from eavesdropping.

> However, encrypting firmware or software update payloads on commodity devices does not constitute an effective cybersecurity defense against

Reviewer: ```suggestion
However, encrypting firmware or software update payloads on commodity devices do not constitute an effective cybersecurity defense against
```

> targeted attacks. Once an attacker gains access to the device, they may still be able to extract the plaintext payload.

Reviewer: ```suggestion
targeted attacks. Once an attacker gains access to a device, they may still be able to extract the plaintext payload.
```

- Evaluation of the required storage and processing resources for the selected profile.
- Support for manifest processing capabilities.

There may be conditions under which switching to a different algorithm profile is not feasible, such as:

- Lack of hardware support (e.g., no crypto acceleration).
- Resource limitations on memory-constrained devices (e.g., insufficient flash or RAM).
- Deployment policy constraints or regulatory compliance requirements.

In such cases, a device management or update orchestration system should take these constraints into account when constructing and distributing manifests.

## Logging and Reporting

Implementations MAY log failures to process a manifest due to unsupported algorithm profiles or unavailable cryptographic functionality. When supported, such events SHOULD be reported using secure mechanisms, such as those described in {{I-D.ietf-suit-report}}, to assist operators in diagnosing update failures or misconfigurations.

# IANA Considerations {#iana}

IANA is requested to create a page for "COSE SUIT Algorithm Profiles" within

> the category for Software Update for the Internet of Things (SUIT). IANA

Reviewer: ```suggestion
the "Software Update for the Internet of Things (SUIT)" registry group. IANA
```

is also requested to create a registry for "COSE SUIT Algorithm Profiles"

> within this page.

Reviewer: ```suggestion
under the same registry group.
```

> The initial content of the "COSE SUIT Algorithm Profiles" registry is:

Reviewer: You may move the initial registry to be positioned right after you define the registration policy and the various status meaning.

## Profile: `suit-sha256-hmac-a128kw-a128ctr`

- Profile: `suit-sha256-hmac-a128kw-a128ctr`
- Status: MANDATORY  
- Digest: -16  
- Auth: 5  
- Key Exchange: -3  
- Encryption: -65534  
- Descriptor Array: `[-16, 5, -3, -65534]`  

> - Reference: Section 3.1

Reviewer: ```suggestion
- Reference: Section 3.1 of THIS_DOCUMENT
```

Reviewer: As this will be in a public registry

## Profile: `suit-sha256-esp256-ecdh-a128ctr`

- Profile: `suit-sha256-esp256-ecdh-a128ctr`
- Status: MANDATORY  
- Digest: -16  
- Auth: -9  
- Key Exchange: -29  
- Encryption: -65534  
- Descriptor Array: `[-16, -9, -29, -65534]`  

> - Reference: Section 3.2

Reviewer: ```suggestion
- Reference: Section 3.2  of THIS_DOCUMENT
```

## Profile: `suit-sha256-ed25519-ecdh-a128ctr`

- Profile: `suit-sha256-ed25519-ecdh-a128ctr`
- Status: MANDATORY  
- Digest: -16  
- Auth: -19  
- Key Exchange: -29  
- Encryption: -65534  
- Descriptor Array: `[-16, -19, -29, -65534]`  

> - Reference: Section 3.3

Reviewer: ```suggestion
- Reference: Section 3.3  of THIS_DOCUMENT
```

## Profile: `suit-sha256-esp256-ecdh-a128gcm`

- Profile: `suit-sha256-esp256-ecdh-a128gcm`
- Status: MANDATORY  
- Digest: -16  
- Auth: -9  
- Key Exchange: -29  
- Encryption: 1  
- Descriptor Array: `[-16, -9, -29, 1]`  

> - Reference: Section 3.4

Reviewer: ```suggestion
- Reference: Section 3.4 of THIS_DOCUMENT
```

## Profile: `suit-sha256-ed25519-ecdh-chacha-poly`

- Profile: `suit-sha256-ed25519-ecdh-chacha-poly`
- Status: MANDATORY  
- Digest: -16  
- Auth: -19  
- Key Exchange: -29  
- Encryption: 24  
- Descriptor Array: `[-16, -19, -29, 24]`  

> - Reference: Section 3.5

Reviewer: ```suggestion
- Reference: Section 3.5 of THIS_DOCUMENT
```

## Profile: `suit-sha256-hsslms-a256kw-a256ctr`

- Profile: `suit-sha256-hsslms-a256kw-a256ctr`
- Status: MANDATORY  
- Digest: -16  
- Auth: -46  
- Key Exchange: -5  
- Encryption: -65532  
- Descriptor Array: `[-16, -46, -5, -65532]`  

> - Reference: Section 3.6

Reviewer: ```suggestion
- Reference: Section 3.6 of THIS_DOCUMENT
```

> 

Reviewer: ```suggestion

> RFC Editor Note: Please replace "THIS_DOCUMENT" with the RFC number assigned to this document.

```

While most profile attributes are self-explanatory, the status field warrants a brief explanation.
This field can take one of three values: MANDATORY, NOT RECOMMENDED, or OPTIONAL.

> - MANDATORY indicates that the profile is mandatory to implement for manifest authors.

Author: I am not sure this change is correct.

Reviewer 2: Mee too, could you make a change to revert it?

- NOT RECOMMENDED means that the profile should generally be avoided in new implementations.
- OPTIONAL suggests that support for the profile is permitted but not required.

> 

Reviewer: ```suggestion
IANA is requested to add a note that mirrors these status values to the registry.

```

Author: Fine for me.

> Adding new profiles or updating the status of existing profiles requires Standards Action.

Reviewer: ```suggestion
Adding new profiles or updating the status of existing profiles requires Standards Action ({{Section 4.9 of !RFC8126}}).
```

As time progresses, algorithm profiles may loose their MANDATORY status. Then, their status will become

> either OPTIONAL or NOT RECOMMENDED for new implementations. Since it may be impossible to update

Reviewer: ```suggestion
either OPTIONAL or NOT RECOMMENDED for new implementations. Likewise, a profile may be transitioned from OPTIONAL to NOT RECOMMENDED. Since it may be impossible to update
```

certain parts of the IoT device firmware in the field, such as first stage bootloaders, support for
all relevant algorithms will almost always be required by authoring tools.

--- back

# Full CDDL {#full-cddl}

> The following CDDL creates a subset of COSE for use with SUIT. Both tagged and untagged messages

Reviewer: ```suggestion
The following CDDL snippet {{!RFC8610}} creates a subset of COSE for use with SUIT. Both tagged and untagged messages
```
